apps/authentication/api/views.py

```diff
@@ -1,22 +1,110 @@
+from rest_framework import status, viewsets
+from rest_framework.permissions import AllowAny, IsAuthenticated
+from rest_framework.response import Response
 
+from apps.authentication.models import Email
 from apps.authentication.models import OnlineUser as User
-from apps.authentication.serializers import UserSerializer
+from apps.authentication.models import Position, SpecialPosition
+from apps.authentication.serializers import (EmailCreateSerializer, EmailReadOnlySerializer,
+                                             EmailUpdateSerializer,
+                                             PositionCreateAndUpdateSerializer,
+                                             PositionReadOnlySerializer, SpecialPositionSerializer,
+                                             UserCreateSerializer, UserReadOnlySerializer,
+                                             UserUpdateSerializer)
 
 
-class UserViewSet(mixins.CreateModelMixin,
-                  mixins.RetrieveModelMixin,
-                  mixins.ListModelMixin,
-                  viewsets.GenericViewSet):
+class UserViewSet(viewsets.ModelViewSet):
     """
     Viewset for User serializer. Supports filtering on 'first_name', 'last_name', 'email'
     """
-
-    queryset = User.objects.all()
-    serializer_class = UserSerializer
     permission_classes = (AllowAny,)
     filterset_fields = ('first_name', 'last_name', 'rfid',)
 
-    def create(self, request, *args, **kwargs):
-        return super().create(request, *args, **kwargs)
+    def get_queryset(self):
+        """
+        Permitted users can view users they have permission for to view.
+        Users can only update/delete their own users.
+        """
+        user = self.request.user
+
+        if self.action in ['list', 'retrieve']:
+            if not user.is_authenticated:
+                return User.objects.none()
+            if user.is_superuser:
+                return User.objects.all()
+            return get_objects_for_user(user, 'authentication.view_onlineuser')
+
+        if self.action in ['list', 'retrieve', 'destroy']:
+            return User.objects.filter(user=user)
+
+        return super().get_queryset()
+
+    def get_serializer_class(self):
+        if self.action == 'create':
+            return UserCreateSerializer
+        if self.action in ['update', 'partial_update']:
+            return UserUpdateSerializer
+        if self.action in ['list', 'retrieve']:
+            return UserReadOnlySerializer
+
+        return super().get_serializer_class()
+
+    def perform_destroy(self, instance: User):
+        """
+        Destroying a user should delete all personal data, and mark the user as inactive.
+        All fields that can be wiped should be wiped, and other fields should be made anonymous.
+        Deleting the user object itself should not be done, as relations like attendance should be preserved.
+        """
+        instance.active = False
+        instance.save()
```

Comment:
Is this method actually doing what the docs are saying?

Reply:
I don't think the API view should handle this. If this is to be done, it should be done as a post_save trigger. And that is outside the scope of this PR (imo).

Reply:
May have been the wrong place to do this, at least the wrong place to write the thoughts.

I did not implement the actual deletion/scrubbing of user data because we have not currently discussed how we should actually do this. We have a lot of user-related data, and we have to find out what we want to delete and what we want to make anonymous.

I think this PR is already a bit too large, and this should be considered a thing to implement in the future.

@sklirg That could be done, do you have any thoughts on how that trigger/signal should be triggered? I suspect you would have to mark the user as deleted or something to make the signal only delete the user-info in specific cases?
The `active` field does not seem like the right thing to use in that case, as I would think that could be set to `False` in other situations.

Did I misunderstand, or is the API-view also the wrong place to trigger account deletion?

Reply:
You can override the delete method on the model.

What I meant is that the API is not the place to implement account deletion functionality. That should be done on the model. The API should in turn call that method for deleting the object, which will use the method from the model.

Reply:
Ah, that sound more like the right way to do it.

I've updated the PR and removed the `destroy` option from the viewset.
It can be re-introduced if user-scrubbing is implemented.